### PR TITLE
Phpstan level 4

### LINKED ISSUE #397: Quoted string parsing doesn't work as expected
As requested in this [PR](https://github.com/microsoft/tolerant-php-parser/pull/396#discussion_r1063150728)

Currently when parsing

```
<?php "'hello world'"';
```

Results in:

```
hello world
```

and not

```
'hello world'
```

same for single and double quotes.

### PR DESCRIPTION
**Pull request by dantleech**
This PR:

- adds support for PHPStan Level 4
- fixes string literal parsing and adds a test for it  (fixes #397)

### CODE REVIEW

File: src/Node/StringLiteral.php [hunk 1]
```diff
@@ -25,17 +25,29 @@ class StringLiteral extends Expression {
...
+                if ($startQuote === '\'') {
+                    return rtrim(substr($value, 1), '\'');
+                }
+
+                if ($startQuote === '"') {
+                    return rtrim(substr($value, 1), '"');
+                }
+            }
+
+            return '';
```
Comment (dantleech) [inline src/Node/StringLiteral.php:50]: this is a feature/bug fix

Comment (roblourens) [reply src/Node/StringLiteral.php:50]: If there isn't an issue for this, could you file one and link this PR to it?

Comment (dantleech) [reply src/Node/StringLiteral.php:50]: sure: https://github.com/microsoft/tolerant-php-parser/issues/397

File: src/Node.php [hunk 2]
```diff
@@ -590,6 +587,7 @@ public function getNamespaceDefinition() {
             ? $this
             : $this->getFirstAncestor(NamespaceDefinition::class, SourceFileNode::class);
 
+        /** @phpstan-ignore-next-line TODO: can this happen? test with framework test cases */
         if ($namespaceDefinition instanceof NamespaceDefinition && !($namespaceDefinition->parent instanceof SourceFileNode)) {
```
Comment (dantleech) [inline src/Node.php:591]: should either ignore this or remove the ignore and update the code and hope that it does never happen.

File: tests/unit/Node/StringLiteralTest.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
+<?php
```
Comment (roblourens) [inline tests/unit/Node/StringLiteralTest.php:1]: Would it make sense to use the pattern in `tests/cases/parser` for this?

Comment (dantleech) [reply tests/unit/Node/StringLiteralTest.php:1]: hm, in what sense? 

keeping this test case but globbing over `../cases/parser/stringLitreal.php`  or something else?

unless I'm missing something there is no way to assert anything with the parser tests as they check the AST structure, while this test is testing a method on the Node.

Comment (roblourens) [reply tests/unit/Node/StringLiteralTest.php:1]: Sorry, I didn't look at it closely enough, yes this makes sense as-is.

**Review by roblourens — APPROVED** [review]

**Comment by roblourens** [comment]
Thanks!

**Pull request merged**